bin/jobserver

```diff
@@ -0,0 +1,386 @@
+import sys
+import socket
+
+import argparse
+import threading
+import os
+import datetime
+import multiprocessing
+from socketserver import ThreadingTCPServer, BaseRequestHandler
+from time import sleep
+
+
+__address__ = "127.0.0.1"
+__port__ = 25721
+__max_jobs__ = 1
+__job_delay__ = 0
+__min_free_mem__ = 2097152
+__max_load__ = None
+__parent_pid__ = None
+__server__ = None
+__primary_server__ = True
+
+
+__job_assign_lock__ = threading.Lock()
+__job_counter_lock__ = threading.Semaphore(10240)
+__job_free_slot_lock__ = threading.Semaphore(__max_jobs__)
+__job_running_lock__ = threading.Semaphore(__max_jobs__)
+__last_active__ = datetime.datetime.utcnow()
+__jobs_status__ = ""
+
+__parent_status__ = ""
+__show_status__ = False
+
+
+class JobServer(BaseRequestHandler):
+    def handle(self):
+        global __job_assign_lock__
+        global __job_counter_lock__
+        global __job_free_slot_lock__
+        global __job_running_lock__
+        global __last_active__
+
+        __job_counter_lock__.acquire()
+        __last_active__ = datetime.datetime.utcnow()
+
+        with __job_assign_lock__:
+            __job_free_slot_lock__.acquire()
+            self._allow_job()
+            __job_running_lock__.acquire()
+            self.request.sendall(b"OK")
+            sleep(__job_delay__)
+
+        while True:
+            try:
+                data = self.request.recv(10)
+            except Exception:
+                break
+            if not data:
+                break
+
+    def finish(self):
+        global __job_counter_lock__
+        global __job_free_slot_lock__
+        global __job_running_lock__
+        global __job_delay__
+
+        __job_running_lock__.release()
+        __job_counter_lock__.release()
+        __job_free_slot_lock__.release()
+
+    def _allow_job(self):
+        global __job_running_lock__
+        global __min_free_mem__
+        global __max_load__
+        global __jobs_status__
+        global __parent_status__
+
+        while True:
+            if __job_running_lock__._value == __max_jobs__:
+                # Allow at least one job to run
+                break
+            if __min_free_mem__ is not None:
+                wait_memory = True
+                with open("/proc/meminfo") as f:
+                    for line in f.readlines():
+                        if line.startswith("MemAvailable:"):
+                            if int(line.split()[1]) < __min_free_mem__:
+                                __jobs_status__ = "Insufficient Free Memory"
+                            else:
+                                wait_memory = False
+                            break
+                if wait_memory:
+                    sleep(1)
+                    continue
+
+            if __max_load__ is not None:
+                if os.getloadavg()[0] > __max_load__:
+                    __jobs_status__ = "System Load Too High"
+                    sleep(1)
+                    continue
+            break
```

Comment:
Instead of an explicit `while True` loop a cleaner approach would be something like
```python
wait = True
while wait:
    wait = False
    if not_enough_memory:
        wait = True
    elif load_too_high:
        wait = True
```